keps/sig-cli/kustomize-reorder-plugins.md

```diff
@@ -0,0 +1,217 @@
+---
+title: Kustomize Resource Plugins Based Reodering
+authors:
+ - "@jbrette"
+owning-sig: sig-cli
+participating-sigs:
+ - sig-apps
+ - sig-api-machinery
+reviewers:
+ - "@monopole"
+ - "@Liujingfang1"
+approvers:
+ - "@monopole"
+editors:
+ - "@jbrette"
+creation-date: 2019-09-01
+last-updated: 2019-09-01
+status: proof of concept
+see-also:
+replaces:
+superseded-by:
+ - n/a
+---
+
+# Kustomize Resource Ordering
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Kustomize orders Resource creation by sorting the Resources it emits based off their type
+or the order used in the original files. The user needs to be able to update that order
+to account for his CRDs and the kubectl operation he intends to perform (apply, delete, patch...)
+
+See [kubernetes-sigs/kustomize#836] for an example.
+See [kubernetes-sigs/kustomize#821] for an example.
+
+## Motivation
+
+Users may need direct control of the Resource create / update / delete ordering in cases were
+sorting by Resource Type is insufficient.
+
+### Goals
+
+- Provide the ability for users to override the order that Kustomize emits Resources
+  - Used by `kubectl apply`
+  - Used by `kubectl delete`
+  
+### Non-Goals
+
+- Ensure certain Resources are *Settled* or *Ready* before other Resources
+- Ensure dependencies between Workloads
+
+## Proposal
+
+Provide a simple mechanism allowing users to override the order that
+Resource operations are applied or deleted. Leverage the plugins feature of kustomize
+in order to let the user override the default sorting plugin.  
+
+```bash
+kustomize build <xxx> --reorder=kubectlapply | kubectl apply -f -
```

Comment:
Are you proposing to use the custom transformer (the order tansformer) with a builtin flag?

Reply:
@Liujingfang1 This is more like a hook. In the current POC, only the names "kubectlapplyordertransformer.yaml" and "kubectldeleteordertransformer.yaml" are hardcoded. Just before outputing the file, kustomize loads the plugin/transformer sort and filter the resmap.